r/tests/testthat/test-dplyr-string-functions.R

```diff
@@ -0,0 +1,172 @@
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+library(dplyr)
+library(stringr)
+
+test_that("sub and gsub", {
+  df <- tibble(x = c("Foo", "bar"))
+
+  for(fixed in c(TRUE, FALSE)) {
+
+    expect_dplyr_equal(
+      input %>%
+        transmute(x = sub("Foo", "baz", x, fixed = fixed)) %>%
+        collect(),
+      df
+    )
+    expect_dplyr_equal(
+      input %>%
+        transmute(x = sub("^B.+", "baz", x, ignore.case = FALSE, fixed = fixed)) %>%
+        collect(),
+      df
+    )
+    expect_dplyr_equal(
+      input %>%
+        transmute(x = sub("Foo", "baz", x, ignore.case = FALSE, fixed = fixed)) %>%
+        collect(),
+      df
+    )
+
+    # the tests below all use ignore.case = TRUE
+    # but base::sub and base::gsub ignore ignore.case = TRUE with a warning when fixed = TRUE
+    # so we can't use expect_dplyr_equal() for the tests below
+    expect_equal(
+      df %>%
+        Table$create() %>%
+        transmute(x = sub("O", "u", x, ignore.case = TRUE, fixed = fixed)) %>%
+        collect(),
+      tibble(x = c("Fuo", "bar"))
+    )
+    expect_equal(
+      df %>%
+        Table$create() %>%
+        transmute(x = gsub("o", "u", x, ignore.case = TRUE, fixed = fixed)) %>%
+        collect(),
+      tibble(x = c("Fuu", "bar"))
+    )
+    expect_equal(
+      df %>%
+        Table$create() %>%
+        transmute(x = sub("^B.+", "baz", x, ignore.case = TRUE, fixed = fixed)) %>%
+        collect(),
+      if (fixed) tibble(x = c("Foo", "bar")) else tibble(x = c("Foo", "baz"))
+    )
+
+  }
+
+})
+
+test_that("str_replace and str_replace_all", {
+  df <- tibble(x = c("Foo", "bar"))
+
+  library(stringr)
+
+  expect_dplyr_equal(
+    input %>%
+      transmute(x = str_replace_all(x, regex("^F"), "baz")) %>%
+      collect(),
+    df
+  )
+  expect_dplyr_equal(
+    input %>%
+      transmute(x = str_replace(x, regex("^f[A-Z]{2}", ignore_case = TRUE), "baz")) %>%
+      collect(),
+    df
+  )
+  expect_dplyr_equal(
+    input %>%
+      transmute(x = str_replace_all(x, fixed("o"), "u")) %>%
+      collect(),
+    df
+  )
+  expect_dplyr_equal(
+    input %>%
+      transmute(x = str_replace(x, fixed("O"), "u")) %>%
+      collect(),
+    df
+  )
+  expect_dplyr_equal(
+    input %>%
+      transmute(x = str_replace(x, fixed("O", ignore_case = TRUE), "u")) %>%
+      collect(),
+    df
+  )
+
+})
+
+test_that("backreferences", {
+  df <- tibble(x = c("Foo", "bar"))
+
+  expect_dplyr_equal(
+    input %>%
+      transmute(desc = sub(
+        "(?:https?|ftp)://([^/\r\n]+)(/[^\r\n]*)?",
+        "path `\\2` on server `\\1`",
+        url
+        )
+      ) %>%
+      collect(),
+    tibble(url = "https://arrow.apache.org/docs/r/")
+  )
+  expect_dplyr_equal(
+    input %>%
+      transmute(x = str_replace(x, regex("^(\\w)o(.*)", ignore_case = TRUE), "\\1\\2p")) %>%
+      collect(),
+    df
+  )
+})
+
+test_that("edge cases", {
```

Comment:
Why are these not `expect_dplyr_equal`?

Reply:
One of them can't use `expect_dplyr_equal()` because it uses `ignore.case = TRUE, fixed = TRUE` which isn't supported in base R. I added a note to explain that, like the note above explaining the same. I'll switch the other example here to use `expect_dplyr_equal()`.

Reply:
Changed in 7aee83c854ac87db17054a7692b35e267e81a0c0